File: python/Problem3.py
# coding:utf-8
"""
The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143 ?
"""
import math
import fractions
import random
import functools


def lcg(n, c, x):
"""
    a pseudorandom number generator (known as Linear Congruential Generator)
"""
    return (x * x + c) % n


class memoize(object):
"""
    a decorator function class to memoize a expensive function.
"""
    def __init__(self, func):
        self._func = func
        self._cache = {}

    def __call__(self, *args):
        if self._cache.has_key(args):
            return self._cache[args]
        res = self._func(*args)
        self._cache[args] = res
        return res


@memoize
def is_prime(n, accuracy=20):
"""
    check if `n` is probable prime using Rabin Miller method.
    to get more reliability, increase `accuracy` parameter.
"""
    n = abs(n)
    if n == 2:
        return True
    if n == 1 or not n & 1:
        return False
    d = n - 1
    while not d & 1:
        d >>= 1
    for i in range(accuracy):
        a = random.randrange(n - 2) + 1
        t = d
        y = pow(a, t, n)
        while t != n - 1 and y != 1 and y != n - 1:
            y = (y * y) % n
            t <<= 1
        if y != n - 1 and not t & 1:
            return False
    return True


def trial_division(n):
"""
    find the smallest divisor of the integer `n` with Trial Division method.
    returns -1 if it is a prime.
"""
    limit = int(math.sqrt(n) + 1)
    for i in range(2, limit):
        if not n % i:
            return i
    return -1


def pollards_rho(n, c=1):
"""
    find the one of divisors of the composite number `n` with Pollard's Rho method.
    returns -1 if it is a prime.
    to get another result, change `c` parameter.
    warning: less reliable for small composite numbers.
"""
    x, y, d = 2, 2, 1
    f = functools.partial(lcg, n, c)
    while d == 1 and n:
        x = f(x)
        y = f(f(y))
        d = fractions.gcd(x - y, n)
        if d == n:
            return -1
    return d


@memoize
def factor(n, accuracy=10):
"""
    factorize the integer `n`.
    returns list of prime factors.
"""
    if is_prime(n):
        return [n]
    if n < 100:
        # fallbacks to trial division for a small integer
        divisor = trial_division(n)
    else:
        # `accuracy` is diverted to the constant of Pollard's Rho
        divisor = pollards_rho(n, accuracy)
    if divisor == -1:
        if accuracy:
            # repeat with another constant
            return factor(n, accuracy - 1)
        return [n]
    return factor(n / divisor) + factor(divisor)


if __name__ == '__main__':
    print(max(factor(600851475143)))

File: javascript/Problem3.js
/*
 * The prime factors of 13195 are 5, 7, 13 and 29.
 * 
 * What is the largest prime factor of the number 600851475143 ?
 */

/**
 * find the greatest common divisor amongst `x` and `y`.
 * each of the arguments must be a natural number.
 */
function gcd(x, y) {
  return y ? gcd(y, x % y) : x
}

/**
 * pseudorandom number generator (known as Linear Congruential Generator).
 */
function lcg(c, n, x) {
  return (x * x + c) % n
}

/**
 * find the smallest divisor of the integer `n` with Trial Division method.
 * returns -1 if it is a prime.
 */
function trialDivision(n) {
  if (n > 2) {
    const limit = n ** 0.5
    for (let i = 2; i <= limit; i++) {
      if (n % i === 0) {
        return i
      }
    }
  }
  return -1
}

/**
 * find the one of divisors of the integer `n` with Pollard's Rho method.
 * returns -1 if it is a probable prime.
 * warning: less reliable for small numbers.
 */
function pollardsRho(n) {
  let x = 2
  let y = 2
  let c = 1
  let d = 1
  // partial application
  let f = lcg.bind(null, c, n)
  while (d === 1 && n) {
    x = f(x)
    y = f(f(y))
    d = gcd(Math.abs(x - y), n)
    if (d === n) {
      return -1
    }
  }
  return d
}

/**
 * a wrapper function to memoize a expensive function.
 * to change the context for function call, bind `this` to the context.
 */
function memoize(func, cache = new Map()) {
  return (...args) => {
    if (!cache.has(args)) {
      cache.set(args, func(...args))
    }
    return cache.get(args)
  }
}

/**
 * factorize the integer `n`.
 * returns an array of prime factors.
 */
function factor(n) {
  // fallbacks to trial division for a small integer
  const divisor = (n < 100) ? trialDivision(n) : pollardsRho(n)
  if (divisor === -1) {
    return [n]
  }
  return factor(n / divisor).concat(factor(divisor))
}
// memoize results
factor = memoize.call(null, factor)

// main entry point
console.log(Math.max(...factor(600851475143)))
